File: packages/redis/src/features/session/lua/create-session.ts
import { z } from 'zod'
import source from './create-session.lua'
import type { LuaScript } from '../../../scripts/runner'

export const createSessionParamsSchema = z.object({
  sessionKey: z.string(),
  refreshKey: z.string(),
  userIdxKey: z.string(),
  sessionJson: z.string(),
  refreshJson: z.string(),
  ttlSeconds: z.number().int().positive(),
  expiresAt: z.number().int().positive(),
  sessionId: z.string(),
  now: z.number().int().positive(),
})

export type CreateSessionParams = z.infer<typeof createSessionParamsSchema>

export const createSessionResultSchema = z.literal(1)
export type CreateSessionResult = z.infer<typeof createSessionResultSchema>

export const createSessionScript: LuaScript<CreateSessionParams, CreateSessionResult> = {
  source,
  buildArgs(params) {
    return {
      keys: [params.sessionKey, params.refreshKey, params.userIdxKey],
      argv: [
        params.sessionJson,
        params.refreshJson,
        params.ttlSeconds,
        params.expiresAt,
        params.sessionId,
        params.now,
      ],
    }
  },
  parseResult: createSessionResultSchema,
}

File: packages/redis/src/features/session/lua/create-session.lua
-- create-session.lua
-- Creates a session atomically:
--   1. SET session data with TTL
--   2. SET refresh-token mapping with TTL
--   3. ZADD session ID to user sessions index (score = expiresAt)
--   4. ZREMRANGEBYSCORE to clean up expired index entries
--
-- KEYS[1]  sessionKey
-- KEYS[2]  refreshKey
-- KEYS[3]  userIdxKey
--
-- ARGV[1]  sessionJson
-- ARGV[2]  refreshJson
-- ARGV[3]  ttl (seconds)
-- ARGV[4]  expiresAt (ms timestamp)
-- ARGV[5]  sessionId
-- ARGV[6]  now (ms timestamp)

local sessionKey  = KEYS[1]
local refreshKey  = KEYS[2]
local userIdxKey  = KEYS[3]

local sessionJson = ARGV[1]
local refreshJson = ARGV[2]
local ttl         = tonumber(ARGV[3])
local expiresAt   = tonumber(ARGV[4])
local sessionId   = ARGV[5]
local now         = tonumber(ARGV[6])

redis.call('SET', sessionKey, sessionJson, 'EX', ttl)
redis.call('SET', refreshKey, refreshJson, 'EX', ttl)
redis.call('ZADD', userIdxKey, expiresAt, sessionId)
redis.call('ZREMRANGEBYSCORE', userIdxKey, '-inf', now)

return 1
